apollo_integration_tests: add library call invoke txs to integration tests

itamar-starkware opened the pull request.

ArniStarkware reviewed (commented):
> @ArniStarkware reviewed 3 of 6 files at r3.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/7133)* status: 3 of 6 files reviewed, all discussions resolved (waiting on @alonh5 and @ayeletstarkware)



<!-- Sent from Reviewable.io -->

ArniStarkware reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/7133)* status: 3 of 6 files reviewed, 2 unresolved discussions (waiting on @alonh5 and @ayeletstarkware)

___
*[`crates/mempool_test_utils/src/starknet_api_test_utils.rs` line 442 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/7133#-OY5kWiw3yJFQ-U_iKkV:-OY5kWiw3yJFQ-U_iKkW:undefined) ([raw file](https://github.com/starkware-libs/sequencer/blob/f9ac53e5110c6f82cafeb9ba2e164f585baabc82/crates/mempool_test_utils/src/starknet_api_test_utils.rs#L442)):*
> ```XML
>             .concat()
>             .into(),
>         );
> ```





_Suggestion:_
```XML
        let library_call_calldata = Calldata(
            [
                vec![test_contract.get_class_hash().0, selector_from_name(inner_fn_name).0],
                inner_fn_args.to_vec(),
            ]
            .concat()
            .into(),
        );
```
___
*[`crates/mempool_test_utils/src/starknet_api_test_utils.rs` line 450 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/7133#-OY5kfbj3bonIO94EEsc:-OY5kfbj3bonIO94EEsd:undefined) ([raw file](https://github.com/starkware-libs/sequencer/blob/f9ac53e5110c6f82cafeb9ba2e164f585baabc82/crates/mempool_test_utils/src/starknet_api_test_utils.rs#L450)):*
> ```XML
> 
>         // Now invoke "test_library_call" on the contract, passing in the constructed
>         // inner_calldata.
> ```





_Suggestion:_
```XML
        // Invoke "test_library_call" on the account contract.
```


<!-- Sent from Reviewable.io -->

itamar-starkware commented:
*[`crates/apollo_integration_tests/src/utils.rs` line 88 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/7133#-OY5mS4i-rjLdr0hyCgR:-OY64vkY9YpEsZN-RiUH:b-ees2b1) ([raw file](https://github.com/starkware-libs/sequencer/blob/f9ac53e5110c6f82cafeb9ba2e164f585baabc82/crates/apollo_integration_tests/src/utils.rs#L88)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

This is a property of `create_custom_invoke_txs_scenario`.
</blockquote></details>

Moved to test_custom_invoke txs.


<!-- Sent from Reviewable.io -->

itamar-starkware commented:
*[`crates/mempool_test_utils/src/starknet_api_test_utils.rs` line 433 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/7133#-OY5ll_E2efvbKEfWCEe:-OY65EY-EwwpAV3j5C6y:b-1emodp) ([raw file](https://github.com/starkware-libs/sequencer/blob/f9ac53e5110c6f82cafeb9ba2e164f585baabc82/crates/mempool_test_utils/src/starknet_api_test_utils.rs#L433)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

As in previous PR.
`test_contract` should not be an argument of this function. The entry point `test_library_call` exists in `TestContract` and we count on it.
</blockquote></details>

My perspective is: creating txs should be versatile, the user should have the ability to decide which test contract he wants to be used. Also he has the responsibility to verify the test contract has the function he wants to call.


<!-- Sent from Reviewable.io -->

itamar-starkware commented:
*[`crates/mempool_test_utils/src/starknet_api_test_utils.rs` line 442 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/7133#-OY5kWiw3yJFQ-U_iKkV:-OY69FEV9FrqzOC8hJNz:bb6un1j) ([raw file](https://github.com/starkware-libs/sequencer/blob/f9ac53e5110c6f82cafeb9ba2e164f585baabc82/crates/mempool_test_utils/src/starknet_api_test_utils.rs#L442)):*
> ```XML
>             .concat()
>             .into(),
>         );
> ```

Done


<!-- Sent from Reviewable.io -->

itamar-starkware commented:
*[`crates/mempool_test_utils/src/starknet_api_test_utils.rs` line 450 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/7133#-OY5kfbj3bonIO94EEsc:-OY69QhC7a1SlpA6tEgZ:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/f9ac53e5110c6f82cafeb9ba2e164f585baabc82/crates/mempool_test_utils/src/starknet_api_test_utils.rs#L450)):*
> ```XML
> 
>         // Now invoke "test_library_call" on the contract, passing in the constructed
>         // inner_calldata.
> ```

Done.


<!-- Sent from Reviewable.io -->

ArniStarkware reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/7133)* status: 3 of 6 files reviewed, 1 unresolved discussion (waiting on @alonh5, @ayeletstarkware, and @itamar-starkware)

___
*[`crates/mempool_test_utils/src/starknet_api_test_utils.rs` line 433 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/7133#-OY5ll_E2efvbKEfWCEe:-OY6oGGjAGPpleKJn6jg:b-chfd3y) ([raw file](https://github.com/starkware-libs/sequencer/blob/f9ac53e5110c6f82cafeb9ba2e164f585baabc82/crates/mempool_test_utils/src/starknet_api_test_utils.rs#L433)):*
<details><summary><i>Previously, itamar-starkware wrote…</i></summary><blockquote>

My perspective is: creating txs should be versatile, the user should have the ability to decide which test contract he wants to be used. Also he has the responsibility to verify the test contract has the function he wants to call.
</blockquote></details>

1. As in the previous PR - I disagree.
2. If we were to eventually go with this approach, please document this assumption on the variable `test_contract`:
```
We assume the contract `test_contract` has an entry point `test_library_call` with a matching signature.
```
This might lead us down a rabbit hole.


<!-- Sent from Reviewable.io -->

itamar-starkware commented:
*[`crates/mempool_test_utils/src/starknet_api_test_utils.rs` line 433 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/7133#-OY5ll_E2efvbKEfWCEe:-OYQ2P3_A5W3t_Q-bZHB:b-fgih3j) ([raw file](https://github.com/starkware-libs/sequencer/blob/f9ac53e5110c6f82cafeb9ba2e164f585baabc82/crates/mempool_test_utils/src/starknet_api_test_utils.rs#L433)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

1. As in the previous PR - I disagree.
2. If we were to eventually go with this approach, please document this assumption on the variable `test_contract`:
```
We assume the contract `test_contract` has an entry point `test_library_call` with a matching signature.
```
This might lead us down a rabbit hole.
</blockquote></details>

test_contract is initialized inside the funcitons, as discussed in previous PR.


<!-- Sent from Reviewable.io -->

ArniStarkware reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @ArniStarkware reviewed 2 of 3 files at r4, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/7133#-:-OYQv97G40LOFxXwqZMG:bnfp4nl)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @alonh5 and @ayeletstarkware)



<!-- Sent from Reviewable.io -->

itamar-starkware commented:
*[`crates/mempool_test_utils/src/starknet_api_test_utils.rs` line 426 at r6](https://reviewable.io/reviews/starkware-libs/sequencer/7133#-OYktYwdCoHOoQR7-NMU:-OZ-FO8q2ZHJ3bL7ss-7:b-hm9ekt) ([raw file](https://github.com/starkware-libs/sequencer/blob/d6dffe457b7b15e2ce7fd195b82da33333bbbf62/crates/mempool_test_utils/src/starknet_api_test_utils.rs#L426)):*
<details><summary><i>Previously, ayeletstarkware (Ayelet Zilber) wrote…</i></summary><blockquote>

please explain (perhaps a comment is needed)
</blockquote></details>

I added a commnet - a call data for library call has a structure so I added an explanation how about this structure.


<!-- Sent from Reviewable.io -->

itamar-starkware commented:
*[`crates/mempool_test_utils/src/starknet_api_test_utils.rs` line 450 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/7133#-OY5kfbj3bonIO94EEsc:-OZ-FbMa2pUp1UjcY_yC:bgov1ex) ([raw file](https://github.com/starkware-libs/sequencer/blob/f9ac53e5110c6f82cafeb9ba2e164f585baabc82/crates/mempool_test_utils/src/starknet_api_test_utils.rs#L450)):*
<details><summary><i>Previously, ayeletstarkware (Ayelet Zilber) wrote…</i></summary><blockquote>

what does it mean?
</blockquote></details>

I wanted to make it clear that we generate and invoke tx of the function test_library_call in this line.


<!-- Sent from Reviewable.io -->

itamar-starkware commented:
*[`crates/mempool_test_utils/src/starknet_api_test_utils.rs` line 450 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/7133#-OY5kfbj3bonIO94EEsc:-OZ3syQbAy_r0vJDuZ2E:bb6un1j) ([raw file](https://github.com/starkware-libs/sequencer/blob/f9ac53e5110c6f82cafeb9ba2e164f585baabc82/crates/mempool_test_utils/src/starknet_api_test_utils.rs#L450)):*
<details><summary><i>Previously, ayeletstarkware (Ayelet Zilber) wrote…</i></summary><blockquote>

remove
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

itamar-starkware commented:
*[`crates/mempool_test_utils/src/starknet_api_test_utils.rs` line 419 at r7](https://reviewable.io/reviews/starkware-libs/sequencer/7133#-OZ3WLp_3b8jxeDZiV0J:-OZ3su1tD3DVxTsrXOB-:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/7382165396bb67e8f22d9c584254a6be3e7c0bec/crates/mempool_test_utils/src/starknet_api_test_utils.rs#L419)):*
> ```XML
>         test_contract: &FeatureContract,
>     ) -> RpcTransaction {
>         // library_call calldata has [class_hash_of_the_contract_to_call, fn_name_selector, fn_args]
> ```

Done.


<!-- Sent from Reviewable.io -->

itamar-starkware commented:
*[`crates/mempool_test_utils/src/starknet_api_test_utils.rs` line 412 at r7](https://reviewable.io/reviews/starkware-libs/sequencer/7133#-OZ3Sa7m3q0p3uJ52Xm4:-OZ3rmWv0jxWMUpxaAvZ:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/7382165396bb67e8f22d9c584254a6be3e7c0bec/crates/mempool_test_utils/src/starknet_api_test_utils.rs#L412)):*
> ```XML
>     }
> 
>     pub fn generate_invoke_tx_library_call(
> ```

Done.


<!-- Sent from Reviewable.io -->

itamar-starkware commented:
*[`crates/apollo_integration_tests/tests/test_custom_syscall_invoke_txs.rs` line 95 at r7](https://reviewable.io/reviews/starkware-libs/sequencer/7133#-OZ3WfYA2aRwSEoUFrc2:-OZ3q7PD1pZz76ItUzP-:boyufrf) ([raw file](https://github.com/starkware-libs/sequencer/blob/7382165396bb67e8f22d9c584254a6be3e7c0bec/crates/apollo_integration_tests/tests/test_custom_syscall_invoke_txs.rs#L95)):*
<details><summary><i>Previously, ayeletstarkware (Ayelet Zilber) wrote…</i></summary><blockquote>

if no arguments needed, why do you pass a vector of felts?
</blockquote></details>

It's actually Cairo behavior. When a library call parses the arguments, it looks for the first element in the vector to determine the number of arguments.
So we need to pass a vector with the value `0` so the library call will know there are no arguemtn in the nested function.


<!-- Sent from Reviewable.io -->

ayeletstarkware reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @ayeletstarkware reviewed 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/7133#-:-OZ9ggkt4Skdudo6nhRJ:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @alonh5 and @itamar-starkware)

___
*[`crates/apollo_integration_tests/tests/test_custom_syscall_invoke_txs.rs` line 95 at r7](https://reviewable.io/reviews/starkware-libs/sequencer/7133#-OZ3WfYA2aRwSEoUFrc2:-OZ9gWLi7mSzXKXhCPfN:b-2v9f3a) ([raw file](https://github.com/starkware-libs/sequencer/blob/7382165396bb67e8f22d9c584254a6be3e7c0bec/crates/apollo_integration_tests/tests/test_custom_syscall_invoke_txs.rs#L95)):*
<details><summary><i>Previously, itamar-starkware wrote…</i></summary><blockquote>

It's actually Cairo behavior. When a library call parses the arguments, it looks for the first element in the vector to determine the number of arguments.
So we need to pass a vector with the value `0` so the library call will know there are no arguemtn in the nested function.
</blockquote></details>

interesting! thanks! 
___
*[`crates/mempool_test_utils/src/starknet_api_test_utils.rs` line 412 at r7](https://reviewable.io/reviews/starkware-libs/sequencer/7133#-OZ3Sa7m3q0p3uJ52Xm4:-OZ9gAi5-VAKe9zyAGtJ:b-jb09jd) ([raw file](https://github.com/starkware-libs/sequencer/blob/7382165396bb67e8f22d9c584254a6be3e7c0bec/crates/mempool_test_utils/src/starknet_api_test_utils.rs#L412)):*
<details><summary><i>Previously, itamar-starkware wrote…</i></summary><blockquote>

Done.
</blockquote></details>

you didn't add `nested`
was it by mistake or you think it's unnecessary?


<!-- Sent from Reviewable.io -->

itamar-starkware commented:
*[`crates/mempool_test_utils/src/starknet_api_test_utils.rs` line 412 at r7](https://reviewable.io/reviews/starkware-libs/sequencer/7133#-OZ3Sa7m3q0p3uJ52Xm4:-OZDcOh23e56UwtLQy7l:b5dhb5n) ([raw file](https://github.com/starkware-libs/sequencer/blob/7382165396bb67e8f22d9c584254a6be3e7c0bec/crates/mempool_test_utils/src/starknet_api_test_utils.rs#L412)):*
<details><summary><i>Previously, ayeletstarkware (Ayelet Zilber) wrote…</i></summary><blockquote>

you didn't add `nested`
was it by mistake or you think it's unnecessary?
</blockquote></details>

On test_custom_syscall_invole_txs, where we build many kinds of transactions, I want to give information that distinguishes them from others.
This is the generic function that generates library invoke tx. I don't want here much information that sounds unique to a specific test.


<!-- Sent from Reviewable.io -->

itamar-starkware reviewed (commented):
> @itamar-starkware reviewed 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/7133)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @alonh5)



<!-- Sent from Reviewable.io -->
